### DISCUSSION TITLE
How to normalize an homography matrix?

### DISCUSSION THREAD

**Discussion by pierrot-lc** [kornia/kornia#2045] [Questions] [answered]
Hi !
I'm using `kornia` to train a neural network that produces homography matricies based on pairs of input images. I'm using normalized homographies (elements in range `[-1, 1]`). Having normalized matricies are more interpretable for me and I can directly penalize the model so that it produces matricies close to the identity homography.

I first tried by training the model to output the homography directly (predicting a 3x3 matrix) which works pretty well:
```py
    def head_direct_homography(self, x: torch.Tensor) -> torch.Tensor:
        """Estimate the homography matrix directly from the logits.

        ----
        Args
            x: Logits extracted from the resnet layers.
                Shape of [batch_size, 3, 3].

        -------
        Returns
            h: Homography matricies.
                Shape of [batch_size, 3, 3].
        """
        # Apply tanh and bias the homography matrix to the identity.
        identity = torch.eye(3, device=x.device) * 1.5
        identity = einops.repeat(identity, "h w -> b h w", b=x.shape[0])
        h = torch.tanh(identity + x)
        return h
```
But I saw that it was better for a model to use the DLT from 4 predicted points instead. So I've tried to use the `kornia.geometry.find_homography_dlt` which, if i'm understanding it correctly, does produce an homography in *pixel-space* and not in the *normal-space*:
```py
    def head_dlt_homography(
        self, x: torch.Tensor, width: int, height: int
    ) -> torch.Tensor:
        """Estimate the homography matrix by using the DLT.

        ----
        Args
            x: Points extracted from the resnet layers.
                Each coordinate is in range `[-1, 1]`.
                Shape of [batch_size, 4, 2].
            width: Width of the original images.
            height: Height of the original images.

        -------
        Returns
            h: Homography matricies.
                Shape of [batch_size, 3, 3].
        """
        # Points in the original image.
        y = torch.FloatTensor([[0, 0], [0, 1], [1, 1], [1, 0]]).to(x.device)
        y = einops.repeat(y, "n c -> b n c", b=x.shape[0])

        # Rescale to match the original pixel space.
        rescale = torch.FloatTensor([[height, width]]).to(x.device)
        y = y * rescale
        x = x * rescale

        # Compute homography in pixel space and convert it to normalized space.
        h = find_homography_dlt(x, y)
        normal_transform = normal_transform_pixel(height, width, device=x.device)
        h = normal_transform @ h

        assert (h.abs() <= 1.1).all(), "Non-normalized homography!"

        return h
```

My question is: how can I normalize an homography that is in pixel-space to the normal-space?

Once I have the batch of homographies, I warp the source images to their destination images and use the L1 Loss to train my model. To warp the images I use the `kornia.geometry.HomographyWarper` that takes as input a batch of normalized homographies.

I saw a `kornia.geometry.normalize_homography` function but it seems to only convert a pixel-space homography from another pixel-space homography (`dsize_src` to `dsize_dst`). From my understanding, I though that by using the `kornia.geometry.normal_transform_pixel` function would work since it produces a matrix that project from the pixel-space to the normal-space.

I may have misunderstood some things, do not hesitate to correct me if I'm wrong somewhere in my explanations.

Thanks a lot for reading my post and I hope I've been clear enough.
Have a good day,
Pierre

**Comment by edgarriba** [comment, marked as answer]
hi! i created a quick colab using instead `get_perspective_transform` and to me the logic seems correct.
https://colab.research.google.com/drive/1PuGz877XhKQy58Kc7TeKCfmbCrwTilJR?usp=sharing

```python
h, w = 320, 480
scale = 1.

x1_pix = torch.tensor([[[0., 0.], [w-1., 0.], [w-1., h-1.], [0., h-1.]]])
x2_pix = x1_pix * scale
print(x1_pix)
print(x2_pix)

#x2_pix_trans_x1_pix = K.geometry.transform.get_perspective_transform(x1_pix, x2_pix)
x2_pix_trans_x1_pix = K.geometry.homography.find_homography_dlt(x1_pix+1e-6, x2_pix+1e-6)
print(x2_pix_trans_x1_pix)

norm_trans_pix= K.geometry.conversions.normal_transform_pixel(h, w)
print(norm_trans_pix)

x2_norm_trans_x1_pix = norm_trans_pix @ x2_pix_trans_x1_pix
print(x2_norm_trans_x1_pix)

# normalize pixels to [-1, 1]
x1_norm = K.geometry.linalg.transform_points(norm_trans_pix, x1_pix)
print(x1_norm)

# transform from x1 in pixels to x2 in pixels
x2_pix = K.geometry.linalg.transform_points(x2_pix_trans_x1_pix, x1_pix)
print(x2_pix)

# transform from x1 in pixels to normalized x2
x2_norm = K.geometry.linalg.transform_points(x2_norm_trans_x1_pix, x1_pix)
print(x2_norm)
```

I also recommend when debugging this problems to name well variables to know the source and destination reference system.

Get perspective transform will solve for the specific case of 4 points and will be more faster that DLT which is mean for N points.

Maybe there is a bug somewhere in `find_homography_dlt` /cc @ducha-aiki

**Reply by pierrot-lc** [reply]
Hi, thank you for your quick response. I wasn't aware of the `get_perspective_transform` function.

I tried with your solution and I got an error for some specific points `x2_pix`. I reproduced the example with a case where the normalization fails:

Input:
```py
h, w = 126, 510
scale = 1.

x1_pix = torch.tensor([[
    [261.6607,  68.6490],
    [195.3000,  49.3307],
    [163.5243,  66.9067],
    [304.5015,  76.4002],
]])
x2_pix = torch.tensor([[
    [0., 0.],
    [w-1., 0.],
    [w-1., h-1.],
    [0., h-1.]
]])
# x2_pix = x1_pix * scale
print(x1_pix)
print(x2_pix)

x2_pix_trans_x1_pix = K.geometry.transform.get_perspective_transform(x1_pix, x2_pix)
print(x2_pix_trans_x1_pix)

norm_trans_pix= K.geometry.conversions.normal_transform_pixel(h, w)
print(norm_trans_pix)

x2_norm_trans_x1_pix = norm_trans_pix @ x2_pix_trans_x1_pix
print(x2_norm_trans_x1_pix)
assert (x2_norm_trans_x1_pix.abs() <= 1.1).all(), "Non-normalized homography!"
```

Output:
```
tensor([[[261.6607,  68.6490],
         [195.3000,  49.3307],
         [163.5243,  66.9067],
         [304.5015,  76.4002]]])
tensor([[[  0.,   0.],
         [509.,   0.],
         [509., 125.],
         [  0., 125.]]])
tensor([[[ 5.6715e+00, -3.1347e+01,  6.6789e+02],
         [ 1.3377e+00, -4.5953e+00, -3.4571e+01],
         [ 1.2240e-02, -5.9601e-02,  1.0000e+00]]])
tensor([[[ 0.0039,  0.0000, -1.0000],
         [ 0.0000,  0.0160, -1.0000],
         [ 0.0000,  0.0000,  1.0000]]])
tensor([[[ 0.0100, -0.0636,  1.6243],
         [ 0.0092, -0.0139, -1.5531],
         [ 0.0122, -0.0596,  1.0000]]])
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
[<ipython-input-13-95200c89bc4d>](https://localhost:8080/#) in <module>
     26 x2_norm_trans_x1_pix = norm_trans_pix @ x2_pix_trans_x1_pix
     27 print(x2_norm_trans_x1_pix)
---> 28 assert (x2_norm_trans_x1_pix.abs() <= 1.1).all(), "Non-normalized homography!"
     29 
     30 # normalize pixels to [-1, 1]

AssertionError: Non-normalized homography!
```

However I have the feeling that the homography has low-range elements (rarely above 2.0 something like that). Maybe my problem is due to some numerical instability?

Thanks again!

**Reply by edgarriba** [reply]
i do not understand that assertion -- if you can convert forth and back points with enough precision i believe it should be fine as unit test

**Reply by pierrot-lc** [reply]
Oh then I think that I misunderstood the whole concept. I though that a normalized homography matrix is supposed to have it's elements in range `[-1, 1]`, but from what I now understand the role of a normalized homography matrix is to be applied to *normalized points* where points have their coordinates in range `[-1, 1]`.

Is that correct?

**Reply by ducha-aiki** [reply]
Yes, that is correct.

**Reply by edgarriba** [reply]
yeah, in any case we should double case about dlt since was throwing nans in some cases respect to the perspective function

**Reply by pierrot-lc** [reply]
Okay now it's clear, thanks a lot!
